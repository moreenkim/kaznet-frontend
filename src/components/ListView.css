.kaznet-table a {
  color: #222;
  font-weight: 700;
}

.kaznet-table a:hover,
.kaznet-table a:focus {
  color: #711c27;
  text-decoration: none;
}

.kaznet-table .dropdown-menu .active-task {
  background: #eee;
  outline: none;
}
