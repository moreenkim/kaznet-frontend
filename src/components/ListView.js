// This component renders a list using Bootstrap 4 tables
import _ from 'lodash';
import React, { Component } from 'react';
import { Table } from 'reactstrap';

export default class ListView extends Component {

  render() {
    return (
      <Table bordered className="kaznet-table">
        <thead>
          <tr>
            {this.props.renderHeaders()}
          </tr>
        </thead>
        <tbody>
          {_.map(this.props.rowsIdArray, this.renderRowById.bind(this))}
        </tbody>
      </Table>
    );
  }

  renderRowById(rowId) {
    return (
      <tr key={rowId}>
        {this.props.renderRow(_.get(this.props.rowsById, rowId))}
      </tr>
    );
  }

}