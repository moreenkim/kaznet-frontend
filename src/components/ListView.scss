@import 'variables.scss';

.kaznet-table a {
  color: #222;
  font-weight: 700;
}

.kaznet-table a:hover,
.kaznet-table a:focus {
  color: $primary-color;
  text-decoration: none;
}

.kaznet-table .dropdown-menu .active-task {
  background: #eee;
  outline: none;
}
