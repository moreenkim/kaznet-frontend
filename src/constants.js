export const API_ENDPOINT = process.env.REACT_APP_KAZNET_ENDPOINT;
export const API_TOKEN = process.env.REACT_APP_API_TOKEN;
export const TASK_ACTIVE = "a";
export const TASK_DEACTIVATED = "b";
export const TASK_EXPIRED = "c";
export const TASK_DRAFT = "d";
export const TASK_SCHEDULED = "s";
export const TASK_ARCHIVED = "e";
export const TASK_STATUSES = [
  TASK_ACTIVE,
  TASK_DEACTIVATED,
  TASK_EXPIRED,
  TASK_DRAFT,
  TASK_SCHEDULED,
  TASK_ARCHIVED
];
