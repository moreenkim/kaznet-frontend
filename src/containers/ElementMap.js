// This component takes a list of strings
// and returns them wrapped in the provided HTML tag
import React, { Component } from "react";
import { DropdownToggle, DropdownMenu, Dropdown } from "reactstrap";
import PropTypes from "prop-types";

export default class ElementMap extends Component {
  static propTypes = {
    filterFields: PropTypes.object
  };

  render() {
    const items = this.props.items;
    // use React.createElement to create the html tag
    // in a dynamic way
    const listItems = items.map((item, index) =>
      React.createElement(
        this.props.HTMLTag,
        { key: index, className: this.props.className },
        [item]
      )
    );

    if (this.props.filterFields != null) {
      /// Check for filterFields prop
      /// If present create a DropDown wrapped inside specified
      /// HTMLTag prop.
      const sortItems = Object.entries(this.props.filterFields).map(
        ([name, dropDownItems], index) => {
          const uniqueKey = 2 * listItems.length;
          return React.createElement(
            this.props.HTMLTag,
            { key: uniqueKey + index, className: this.props.className },
            [
              <Dropdown
                key={uniqueKey * 2}
                isOpen={this.props.isOpen}
                toggle={e => this.props.handleChange(e)}
              >
                <DropdownToggle caret tag="span" className="nav-link">
                  {name}
                </DropdownToggle>
                <DropdownMenu>{dropDownItems}</DropdownMenu>
              </Dropdown>
            ]
          );
        }
      );

      /// Modify listItems to Include Filter Fields in Specified Position in
      /// this.props.filterItemPositions
      for (var i = 0; i < this.props.filterItemPositions.length; i++) {
        listItems.splice(this.props.filterItemPositions[i], 0, sortItems[i]);
      }
    }

    return listItems;
  }
}
