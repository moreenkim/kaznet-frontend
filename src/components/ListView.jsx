// This component renders a list using Bootstrap 4 tables
import _ from 'lodash';
import React, { Component } from 'react';
import { Table, Pagination, PaginationItem } from 'reactstrap';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

import * as constants from '../constants';
import ExportModal from './ExportModal';

import './ListView.scss';

export default class ListView extends Component {
  renderRowById(rowId) {
    const { renderRow, rowsById, downloadModalHandler, setUserDetails } = this.props;
    return (
      <tr key={rowId}>{renderRow(_.get(rowsById, rowId), downloadModalHandler, setUserDetails)}</tr>
    );
  }

  renderRowObject(rowObject) {
    const { renderRow, downloadModalHandler, setUserDetails } = this.props;
    return <tr key={rowObject.id}>{renderRow(rowObject, downloadModalHandler, setUserDetails)}</tr>;
  }

  renderPagination() {
    const {
      endpoint,
      pageLinks,
      searchVal,
      firstPage,
      lastPage,
      totalPages,
      taskStatus,
      hasTask,
      currentPage
    } = this.props;
    return (
      <Pagination aria-label="Kaznet listview">
        <PaginationItem disabled={!pageLinks.first}>
          <Link
            to={
              pageLinks.first
                ? `/${endpoint}/?search=${
                    !searchVal || searchVal === undefined ? '' : searchVal
                  }&page=${!firstPage || typeof firstPage !== 'number' ? 1 : firstPage}${
                    taskStatus ? `&status=${taskStatus}` : ''
                  }${hasTask ? `&has_task=${hasTask}` : ''}`
                : '#'
            }
            className="page-link"
            aria-label="First"
          >
            First
          </Link>
        </PaginationItem>
        <PaginationItem disabled={!pageLinks.prev}>
          <Link
            to={
              pageLinks.prev
                ? `/${endpoint}/?search=${!searchVal ? '' : searchVal}&page=${currentPage - 1}${
                    taskStatus ? `&status=${taskStatus}` : ''
                  }${hasTask ? `&has_task=${hasTask}` : ''}`
                : '#'
            }
            className="page-link"
            aria-label="Previous"
          >
            &laquo;
          </Link>
        </PaginationItem>
        <PaginationItem disabled={!pageLinks.next}>
          <Link
            to={
              pageLinks.next
                ? `/${endpoint}/?search=${!searchVal ? '' : searchVal}&page=${currentPage + 1}${
                    taskStatus ? `&status=${taskStatus}` : ''
                  }${hasTask ? `&has_task=${hasTask}` : ''}`
                : '#'
            }
            className="page-link"
            aria-label="Next"
          >
            &raquo;
          </Link>
        </PaginationItem>
        <PaginationItem disabled={!pageLinks.last}>
          <Link
            to={
              pageLinks.last
                ? `/${endpoint}/?search=${!searchVal ? '' : searchVal}&page=${
                    !lastPage || lastPage === undefined ? totalPages : lastPage
                  }${taskStatus ? `&status=${taskStatus}` : ''}${
                    hasTask ? `&has_task=${hasTask}` : ''
                  }`
                : '#'
            }
            className="page-link"
            aria-label="Last"
          >
            Last
          </Link>
        </PaginationItem>
      </Pagination>
    );
  }

  render() {
    const {
      sortField,
      downloadModalHandler,
      modalState,
      className,
      onFormSubmit,
      userName,
      isFormPage,
      renderHeaders,
      pageLinks,
      endpoint,
      searchVal,
      firstPage,
      rowsById,
      totalPages,
      rowsIdArray,
      sortOrder
    } = this.props;

    if (sortField) {
      const rowsArray = _.values(rowsById);
      this.sortedRowsArray = _.orderBy(
        rowsArray,
        e => {
          return e.attributes[sortField];
        },
        [sortOrder]
      );
    }

    return (
      <div>
        {downloadModalHandler && (
          <ExportModal
            modalState={modalState}
            downloadModalHandler={downloadModalHandler}
            className={className}
            onFormSubmit={onFormSubmit}
            name={userName}
          />
        )}

        <Table bordered className="kaznet-table">
          <thead>
            <tr>
              {isFormPage
                ? renderHeaders(
                    pageLinks.first
                      ? `/${endpoint}/?search=${
                          !searchVal || searchVal === undefined ? '' : searchVal
                        }&page=${!firstPage || typeof firstPage !== 'number' ? 1 : firstPage}`
                      : '#'
                  )
                : renderHeaders()}
            </tr>
          </thead>
          <tbody>
            {sortField
              ? _.map(this.sortedRowsArray, this.renderRowObject.bind(this))
              : _.map(rowsIdArray, this.renderRowById.bind(this))}
          </tbody>
        </Table>
        {totalPages > 1 ? this.renderPagination() : null}
      </div>
    );
  }
}

ListView.propTypes = {
  sortField: PropTypes.string,
  sortOrder: PropTypes.string,
  rowsById: PropTypes.shape({}).isRequired,
  downloadModalHandler: PropTypes.func.isRequired,
  modalState: PropTypes.bool.isRequired,
  className: PropTypes.string.isRequired,
  onFormSubmit: PropTypes.func.isRequired,
  userName: PropTypes.string.isRequired,
  isFormPage: PropTypes.bool.isRequired,
  renderHeaders: PropTypes.func.isRequired,
  renderRow: PropTypes.func.isRequired,
  setUserDetails: PropTypes.func.isRequired,
  pageLinks: PropTypes.shape({
    first: PropTypes.string,
    prev: PropTypes.string,
    last: PropTypes.string,
    next: PropTypes.string
  }).isRequired,
  endpoint: PropTypes.string.isRequired,
  searchVal: PropTypes.string.isRequired,
  taskStatus: PropTypes.string.isRequired,
  hasTask: PropTypes.bool.isRequired,
  currentPage: PropTypes.number,
  firstPage: PropTypes.number,
  lastPage: PropTypes.number,
  totalPages: PropTypes.number,
  rowsIdArray: PropTypes.arrayOf().isRequired
};

ListView.defaultProps = {
  sortOrder: constants.SORT_ASC,
  sortField: null,
  firstPage: 1,
  lastPage: 1,
  totalPages: 1,
  currentPage: 1
};
