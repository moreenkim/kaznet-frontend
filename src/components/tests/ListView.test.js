// Test ListView
import React from "react";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";

import ErrorBoundary from "../ErrorBoundary";
import ElementMap from "../../containers/ElementMap";
import ListView from "../ListView";
import * as constants from "../../constants";

import { Router } from "react-router";
import createBrowserHistory from "history/createBrowserHistory";
const history = createBrowserHistory();

describe("components/ListView", () => {
  it("renders without crashing", () => {
    const rowsById = { "0": "Bob", "1": "Jane" };
    const rowsIdArray = ["0", "1"];

    const renderHeader = function() {
      const headerItems = ["Name"];

      return <ElementMap items={headerItems} HTMLTag="th" />;
    };

    const renderRow = function(row) {
      const rowItems = [row];
      return <ElementMap items={rowItems} HTMLTag="td" />;
    };
    const endpoint = `${constants.API_ENDPOINT}`;
    const currentPage = 1;
    const totalPages = 2;
    const firstPage = 1;
    const lastPage = 2;
    const pageLinks = {
      first: "http://localhost:8000/api/v1/clients/?page=1",
      last: "http://localhost:8000/api/v1/clients/?page=2",
      next: "http://localhost:8000/api/v1/clients/?page=2",
      prev: null
    };

    shallow(
      <Router history={history}>
        <ListView
          endpoint={endpoint}
          renderHeaders={renderHeader}
          rowsIdArray={rowsIdArray}
          rowsById={rowsById}
          renderRow={renderRow}
          firstPage={firstPage}
          lastpage={lastPage}
          pageLinks={pageLinks}
          totalPages={totalPages}
          currentPage={currentPage}
          searchVal={"test"}
          taskStatus={"2"}
        />
      </Router>
    );
  });

  it("renders a table", () => {
    const rowsById = { "0": "Bob", "1": "Jane" };
    const rowsIdArray = ["0", "1"];

    const renderHeader = function() {
      const headerItems = ["Name"];
      return <ElementMap items={headerItems} HTMLTag="th" />;
    };

    const renderRow = function(row) {
      const rowItems = [row];
      return <ElementMap items={rowItems} HTMLTag="td" />;
    };

    const endpoint = `${constants.API_ENDPOINT}`;
    const currentPage = 1;
    const totalPages = 2;
    const firstPage = 1;
    const lastPage = 2;
    const pageLinks = {
      first: "http://localhost:8000/api/v1/clients/?page=1",
      last: "http://localhost:8000/api/v1/clients/?page=2",
      next: "http://localhost:8000/api/v1/clients/?page=2",
      prev: null
    };

    const wrapper = mount(
      <Router history={history}>
        <ErrorBoundary>
          <ListView
            endpoint={endpoint}
            renderHeaders={renderHeader}
            rowsIdArray={rowsIdArray}
            rowsById={rowsById}
            renderRow={renderRow}
            firstPage={firstPage}
            lastpage={lastPage}
            pageLinks={pageLinks}
            totalPages={totalPages}
            currentPage={currentPage}
            searchVal={"test"}
            taskStatus={"2"}
          />
        </ErrorBoundary>
      </Router>
    );

    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper.unmount();
  });

  it("renders pagination when totalPages > 1", () => {
    const rowsById = { "0": "Bob", "1": "Jane" };
    const rowsIdArray = ["0", "1"];

    const renderHeader = function() {
      const headerItems = ["Name"];
      return <ElementMap items={headerItems} HTMLTag="th" />;
    };

    const renderRow = function(row) {
      const rowItems = [row];
      return <ElementMap items={rowItems} HTMLTag="td" />;
    };

    const endpoint = `${constants.API_ENDPOINT}`;
    const currentPage = 1;
    const totalPages = 2;
    const firstPage = 1;
    const lastPage = 2;
    const pageLinks = {
      first: "http://localhost:8000/api/v1/clients/?page=1",
      last: "http://localhost:8000/api/v1/clients/?page=2",
      next: "http://localhost:8000/api/v1/clients/?page=2",
      prev: null
    };

    const wrapper = mount(
      <Router history={history}>
        <ListView
          endpoint={endpoint}
          renderHeaders={renderHeader}
          rowsIdArray={rowsIdArray}
          rowsById={rowsById}
          renderRow={renderRow}
          firstPage={firstPage}
          lastpage={lastPage}
          pageLinks={pageLinks}
          totalPages={totalPages}
          currentPage={currentPage}
          searchVal={"test"}
          taskStatus={"2"}
        />
      </Router>
    );

    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper.unmount();
  });

  it("does not render pagination when totalPages <= 1", () => {
    const rowsById = { "0": "Bob", "1": "Jane" };
    const rowsIdArray = ["0", "1"];

    const renderHeader = function() {
      const headerItems = ["Name"];
      return <ElementMap items={headerItems} HTMLTag="th" />;
    };

    const renderRow = function(row) {
      const rowItems = [row];
      return <ElementMap items={rowItems} HTMLTag="td" />;
    };

    const endpoint = `${constants.API_ENDPOINT}`;
    const currentPage = 1;
    const totalPages = 1;
    const firstPage = 1;
    const lastPage = 1;
    const pageLinks = {
      first: "http://localhost:8000/api/v1/clients/?page=1",
      last: null,
      next: null,
      prev: null
    };

    const wrapper = mount(
      <Router history={history}>
        <ListView
          endpoint={endpoint}
          renderHeaders={renderHeader}
          rowsIdArray={rowsIdArray}
          rowsById={rowsById}
          renderRow={renderRow}
          firstPage={firstPage}
          lastpage={lastPage}
          pageLinks={pageLinks}
          totalPages={totalPages}
          currentPage={currentPage}
        />
      </Router>
    );

    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper.unmount();
  });
});
