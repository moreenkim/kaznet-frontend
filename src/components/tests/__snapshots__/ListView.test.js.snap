// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ListView renders a table 1`] = `
<ErrorBoundary>
  <ListView
    renderHeaders={[Function]}
    renderRow={[Function]}
    rowsById={
      Object {
        "0": "Bob",
        "1": "Jane",
      }
    }
    rowsIdArray={
      Array [
        "0",
        "1",
      ]
    }
  >
    <Table
      bordered={true}
      className="kaznet-table"
      responsiveTag="div"
      tag="table"
    >
      <table
        className="kaznet-table table table-bordered"
      >
        <thead>
          <tr>
            <ElementMap
              HTMLTag="th"
              items={
                Array [
                  "Name",
                ]
              }
            >
              <th
                key="0"
              >
                Name
              </th>
            </ElementMap>
          </tr>
        </thead>
        <tbody>
          <tr
            key="0"
          >
            <ElementMap
              HTMLTag="td"
              items={
                Array [
                  "Bob",
                ]
              }
            >
              <td
                key="0"
              >
                Bob
              </td>
            </ElementMap>
          </tr>
          <tr
            key="1"
          >
            <ElementMap
              HTMLTag="td"
              items={
                Array [
                  "Jane",
                ]
              }
            >
              <td
                key="0"
              >
                Jane
              </td>
            </ElementMap>
          </tr>
        </tbody>
      </table>
    </Table>
  </ListView>
</ErrorBoundary>
`;
